code/controllers/subsystem/spatial_gridmap.dm

```diff
@@ -358,56 +411,66 @@ SUBSYSTEM_DEF(spatial_grid)
+				GRID_CELL_REMOVE(intersecting_cell.hearing_contents, old_target_contents[SPATIAL_GRID_CONTENTS_TYPE_HEARING])
+				SEND_SIGNAL(intersecting_cell, SPATIAL_GRID_CELL_EXITED(SPATIAL_GRID_CONTENTS_TYPE_HEARING), old_target_contents[SPATIAL_GRID_CONTENTS_TYPE_HEARING])
+
+			if(SPATIAL_GRID_CONTENTS_TYPE_ATMOS)
+				GRID_CELL_REMOVE(intersecting_cell.atmos_contents, old_target)
+				SEND_SIGNAL(intersecting_cell, SPATIAL_GRID_CELL_EXITED(SPATIAL_GRID_CONTENTS_TYPE_ATMOS), old_target)
 
 	return TRUE
 
 ///acts like exit_cell() but only removes the target from the specified type of grid cell contents list
-/datum/controller/subsystem/spatial_grid/proc/remove_single_contents_type(atom/movable/old_target, turf/target_turf, exclusive_type)
-	if(!target_turf || !old_target?.important_recursive_contents || !exclusive_type)
-		stack_trace("/datum/controller/subsystem/spatial_grid/proc/remove_single_contents_type() was given null arguments or a new_target without important_recursive_contents!")
+/datum/controller/subsystem/spatial_grid/proc/remove_single_type(atom/movable/old_target, turf/target_turf, exclusive_type)
+	if(!target_turf || !exclusive_type || !old_target.spatial_grid_key)
+		stack_trace("/datum/controller/subsystem/spatial_grid/proc/remove_single_type() was given null arguments or an old_target that doesn't use the spatial grid!")
 		return FALSE
 
-	if(!old_target.important_recursive_contents[exclusive_type])
+	if(!(exclusive_type in spatial_grid_categories[old_target.spatial_grid_key]))
 		return FALSE
 
-	var/x_index = ROUND_UP(target_turf.x / SPATIAL_GRID_CELLSIZE)
-	var/y_index = ROUND_UP(target_turf.y / SPATIAL_GRID_CELLSIZE)
+	var/x_index = GET_SPATIAL_INDEX(target_turf.x)
+	var/y_index = GET_SPATIAL_INDEX(target_turf.y)
 	var/z_index = target_turf.z
 
 	var/datum/spatial_grid_cell/intersecting_cell = grids_by_z_level[z_index][y_index][x_index]
 
 	switch(exclusive_type)
 		if(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS)
-			GRID_CELL_REMOVE(intersecting_cell.client_contents, old_target.important_recursive_contents[SPATIAL_GRID_CONTENTS_TYPE_CLIENTS])
+			var/list/old_target_contents = old_target.important_recursive_contents //cache for sanic speeds (lists are references anyways)
+			GRID_CELL_REMOVE(intersecting_cell.client_contents, old_target_contents[SPATIAL_GRID_CONTENTS_TYPE_CLIENTS])
+			SEND_SIGNAL(intersecting_cell, SPATIAL_GRID_CELL_EXITED(exclusive_type), old_target_contents[SPATIAL_GRID_CONTENTS_TYPE_CLIENTS])
 
 		if(SPATIAL_GRID_CONTENTS_TYPE_HEARING)
-			GRID_CELL_REMOVE(intersecting_cell.hearing_contents, old_target.important_recursive_contents[SPATIAL_GRID_CONTENTS_TYPE_HEARING])
+			var/list/old_target_contents = old_target.important_recursive_contents
+			GRID_CELL_REMOVE(intersecting_cell.hearing_contents, old_target_contents[SPATIAL_GRID_CONTENTS_TYPE_HEARING])
+			SEND_SIGNAL(intersecting_cell, SPATIAL_GRID_CELL_EXITED(exclusive_type), old_target.important_recursive_contents[SPATIAL_GRID_CONTENTS_TYPE_HEARING])
```

Comment:
you dont use the old_target_contents list for this signal and the one above

Reply:
thanks boss

Reply:
One above?